### PR TITLE
core: assign block names in region order

### LINKED ISSUE #5567: core: Match MLIR block name logic exactly
The block names are currently chosen in order of them being printed in the IR, including when printed as successors, which results in them having different orders to MLIR, where they are printed in the order they're in in the region.

It would be good to have near-0 effort when using tests from MLIR in xDSL, which is in conflict with inconsistencies in the textual format.

As noted in #5563, there's a simple mapping between the two in many cases:

```
// copied from mlir/test/Conversion/PDLToPDLInterp/pdl-to-pdl-interp-matcher.mlir
// basic block numbering is slightly different, the following perl one-liner gets us far:
// `perl -pe 's/\bbb(\d+)/"bb".($1==2 ? 0 : $1>2 ? $1-1 : $1)/ge'` (replace bb2 with bb0, and decrement all bb numbers > 2)
// The argument list of functions also contains an extra space in xDSL compared to MLIR.
```

This is due to the blocks being inserted by a lowering from regions usually, which apply similar logic. When changing the logic in xDSL to match MLIR, a similar script is likely useful to update the tests.

### PR DESCRIPTION
**Pull request by Sanjays2402**
Block names are now assigned from their position in the containing region before any successor is printed, so traversal order cannot change numeric IDs. This matches MLIR, adds an out-of-order-successor regression, and updates affected expected output, including the copied PDL matcher test that no longer needs its renumbering workaround.

Closes #5567.

### CODE REVIEW

**Comment by superlopuh** [comment]
Happy to review once merge conflicts are resolved. Please remove the "PR template" bit from your description.

File: xdsl/printer.py [hunk 1]
```diff
@@ -177,6 +177,22 @@ def print_ssa_value(self, value: SSAValue) -> str:
...
+            curr_ind = self.block_names.get(block.name_hint, 0)
+            suffix = f"_{curr_ind}" if curr_ind != 0 else ""
+            name = f"{block.name_hint}{suffix}"
+            self.block_names[block.name_hint] = curr_ind + 1
+        elif block_id is not None:
+            name = f"bb{block_id}"
+        else:
+            name = f"bb{self._get_new_valid_block_id()}"
+        self._blocks[block] = name
+        return name
```
Comment (superlopuh) [inline xdsl/printer.py]: Let's change this a tiny bit while we're here:
```suggestion
        try:
            name = self._blocks[block]
        except KeyError:
	        if block.name_hint:
	            curr_ind = self.block_names.get(block.name_hint, 0)
	            suffix = f"_{curr_ind}" if curr_ind != 0 else ""
	            name = f"{block.name_hint}{suffix}"
	            self.block_names[block.name_hint] = curr_ind + 1
	        elif block_id is not None:
	            name = f"bb{block_id}"
	        else:
	            name = f"bb{self._get_new_valid_block_id()}"
	        self._blocks[block] = name
        return name
```

As a matter of style, we prefer "ask for forgiveness, not permission", in this case this avoids looking up `block` in the hashmap twice in the common case of the block being in there.

Comment (Sanjays2402) [reply xdsl/printer.py]: Updated `_get_block_name` to use `try`/`except KeyError`, so the cached path performs a single dictionary lookup. I also rebased onto current `main`; the branch is now conflict-free. The focused printer/rewriter suite (429 tests), Ruff, formatting, and the relevant local FileCheck pipelines pass, and the full CI matrix is running.

**Comment by Sanjays2402** [comment]
Done — I removed the “PR Template” heading and rebased onto current `main`. The `memref_streamify` conflict preserves the upstream stream argument names together with this PR’s region-order block labels. GitHub now reports the PR as mergeable; CI is running on the rebased head.

**Review by superlopuh — APPROVED** [review]
@math-fehr @alexarice, WDYT? I had a quick check locally with mlir-opt and it looks to be consistent with how MLIR names blocks.

File: xdsl/printer.py [hunk 2]
```diff
@@ -257,6 +263,11 @@ def print_region(
         * If `print_empty_block` is False, empty entry blocks are not printed.
         * If `print_block_terminators` is False, the block terminators are not printed.
         """
+        # Match MLIR by assigning block names in region order before successors are
+        # printed.
+        for block_id, block in enumerate(region.blocks):
```
Comment (alexarice) [inline xdsl/printer.py]: I'm not sure I understand why the block_id part is needed here.

**Review by alexarice — APPROVED** [review]
Seems reasonable to me apart from one comment. This will obviously cause many downstream breakages

**Comment by Sanjays2402** [comment]
@alexarice `block_id` preserves the **absolute position in the region**, including positions occupied by blocks with name hints. For example, if position 0 is hinted as `entry`, the next unnamed block at position 1 must be `bb1`; using `_get_new_valid_block_id()` there would incorrectly produce `bb0` because the hinted block does not advance that counter.

I added that case as a regression in `c5baf05a`. A mutation removing the explicit ID changed the expected `bb1` to `bb0`, as above. Final verification: **87 printer tests passed**, with Ruff and formatting clean.

**Comment by alexarice** [comment]
I think mlir doesn't use name hints at all, so realistically we have a choice whether we use absolute position as the index. The method here makes sure that blocks inside regions start counting from 0 again, which probably should have been happening before, and I think this is a bug in `enter_scope`.

The only think I don't really like about this solution is that it feels like there's two parallel ways of assigning names to blocks. If we print a region, then all the blocks get assigned the name given by the index in the region (and then every call to print block name after that just uses the dictionary lookup). If we just print an individual block name, then it gets given a name by `_get_new_valid_block_id`.

**Comment by superlopuh** [comment]
In a way it feels like there could be two functions, one to populate the block names and one to fetch them? It feels like this would make the logic clearer. It also feels like `block_index` would be a better name than `block_id` for what it does.

**Comment by superlopuh** [comment]
Printing a block on its own, outside a region, is pretty rare, right? I think it should be OK to special case that and have two places where we populate block names, one in region printing (how it's done in this PR), and one in printing a block.

File: xdsl/printer.py (hunk 2 above)
Comment (Sanjays2402) [reply xdsl/printer.py]: The index is only needed during region prepopulation so names follow region order even when successor references are printed first. `d10e09e7` separates `_populate_block_name(block, block_index)` from `_get_block_name(block)`, and the lookup path now has no index parameter. The focused printer suite passes 87 tests and Pyright reports no diagnostics.

File: xdsl/printer.py [hunk 3]
```diff
@@ -177,6 +177,32 @@ def print_ssa_value(self, value: SSAValue) -> str:
...
+                name = f"bb{self._get_new_valid_block_id()}"
+            self._blocks[block] = name
+
+    def _get_block_name(self, block: Block) -> str:
+        """Fetch a block name, assigning one on a cache miss."""
+        try:
+            return self._blocks[block]
+        except KeyError:
+            self._populate_block_name(block)
+            return self._blocks[block]
```
Comment (superlopuh) [inline xdsl/printer.py:174]: No, that's the point, the block names should always be populated first, and then they should be already populated by the time the block is printed. This way we can be consistent in a guaranteed way. Does this make sense?
```suggestion
        return self._blocks[block]
```

I would probably also add an assert in the function above
```py
assert block not in self._blocks
```

The one thing to watch out for will be printing a single block but I expect that if you make this change you'll see the error immediately and see how to fix it.

Comment (Sanjays2402) [reply xdsl/printer.py:174]: Done. `_get_block_name` is just the dict lookup now, and `_populate_block_name` asserts the block isnt already named. The one caller that hits an unpopulated block is printing a block on its own outside a region, so I populate there before printing the name. Standalone still prints fine (`^bb0`, or the hint if it has one). 87 printer tests green, pyright clean.

**Review by Samielakkad — COMMENTED** [review]
I think the failing tests point to one remaining edge case in the new block-name population model.

The Python 3.12 failure is:

```text
tests/irdl/test_declarative_assembly_format.py::test_ref_directives
KeyError in xdsl/printer.py:_get_block_name
```

`print_region` now pre-populates names for blocks in region order, which makes sense for normal region printing. But `test_ref_directives` reaches `printer.print_block_name(self.get(op))` through the declarative assembly format ref directive, and that referenced block is not in `_blocks` yet. Since `_get_block_name` now only fetches already-populated names, the ref directive path raises `KeyError`.

So I think one of these needs to happen:

- keep a cache-miss fallback for `print_block_name` / `_get_block_name`, or
- make the declarative assembly format ref-directive path explicitly populate referenced blocks before printing them.

The latter may preserve the new “region order first” invariant better, but either way the current PR still has a non-region-printing path that calls `print_block_name` before prepopulation.

**Comment by superlopuh** [comment]
OK, in this case my last suggestion was not good, I'm not sure how to resolve it. Let's revert the last commit, then should be good to merge.

**Comment by superlopuh** [comment]
I still think `block_index` is the better name, for the argument, though

**Comment by superlopuh** [comment]
Thinking about it more, I think the only change that would be necessary is just reverting to this, the other structure from the last commit is still an improvement:

```py
        """Fetch a block name, assigning one on a cache miss."""
        try:
            return self._blocks[block]
        except KeyError:
            self._populate_block_name(block)
            return self._blocks[block]
```

**Comment by Sanjays2402** [comment]
Reverted _get_block_name to the try/except form as you asked — kept the rest of the changes from the last commit intact, just that method body went back to the cache-miss-on-KeyError structure. Printer tests (87) and the broader block/printer set (332) pass, and the parser-printer filecheck suite is 26/26. Pushed.

**Comment by superlopuh** [comment]
@Sanjays2402  Can you please make sure the tests pass locally? Looking at the error online it's not immediately obvious to me where the bug is.

**Comment by superlopuh** [comment]
@Sanjays2402 when will you have time to fix the failing tests?

**Comment by Sanjays2402** [comment]
@superlopuh Let me check them shortly

**Comment by Sanjays2402** [comment]
@superlopuh Found it. The failure wasn't in the tests, it was `tests-marimo`: `docs/marimo/xdsl_introduction.py` reuses one `Printer` to print the same module twice, so `print_region` hit `_populate_block_name` a second time on blocks that were already named and tripped the new assert.

Fixed by skipping already-named blocks in the region prepass, so a reused printer keeps the names it assigned the first time:

```python
for block_index, block in enumerate(region.blocks):
    if block not in self._blocks:
        self._populate_block_name(block, block_index)
```

Added `test_print_region_twice_with_same_printer` which asserts the two printings are identical; reverting the guard makes it fail with the same AssertionError from CI. `tests/test_printer.py` is 88 passed, and the wider suite is 4906 passed with two failures (`test_multiverse`, `test_run_tblgen_to_py`) that fail identically on the unmodified branch here. Pushed as f6a8af3.

**Review by superlopuh — APPROVED** [review]
I like this. I'm not 100% sure about the details, but the majority of the diff is definitely an improvement IMO.

**Comment by superlopuh** [comment]
@alexarice  if you're still happy with this I would merge it. @Sanjays2402 can you please merge main?

File: xdsl/printer.py [hunk 4]
```diff
@@ -177,6 +177,30 @@ def print_ssa_value(self, value: SSAValue) -> str:
...
+
     def print_block_name(self, block: Block) -> str:
         """
         Print a block name in the printer. This assigns a name to the block if the block
```
Comment (alexarice) [inline xdsl/printer.py:178]: This docstring is no longer accurate

File: xdsl/printer.py [hunk 5]
```diff
@@ -215,6 +228,10 @@ def print_block(
 
         if print_block_args:
             self._print_new_line()
+            # A block printed on its own, outside a region, won't have been
+            # populated yet; region printing populates all names up front.
+            if block not in self._blocks:
+                self._populate_block_name(block)
```
Comment (alexarice) [inline xdsl/printer.py:206]: This feels like it could just be in `print_block_name`, though it feels like the review comments are going in circles so I'm happy with whatever is easiest

**Comment by Sanjays2402** [comment]
Merged main — `c1ea305`, clean merge with no conflicts. CI is green (24 passing, `deploy` skipped) and the PR is showing mergeable.

File: xdsl/printer.py (hunk 4 above)
Comment (superlopuh) [reply xdsl/printer.py:178]: Can you please make a suggestion for how to improve it?

Comment (alexarice) [reply xdsl/printer.py:178]: nvm I misread it it is fine

File: xdsl/printer.py (hunk 5 above)
Comment (alexarice) [reply xdsl/printer.py:206]: Ok, my confusion is actually here, it seems that `print_block_name` *does* already populate the name if it's not set, so I don't see why we check again here

Comment (superlopuh) [reply xdsl/printer.py:206]: At this point in the review I thought we could merge just to get the big diff in, I'd be happy to review a PR that simplifies the mechanism
